internal/cpp_doc_generator/developer_guide_logging.rst

```diff
@@ -0,0 +1,304 @@
+Logging
+-------
+
+In this section, the log mechanisms of iceoryx2 are described in detail. We
+begin by analyzing the structure of log messages and how they can be utilized
+when debugging a process. Next, we describe the error-handling strategy, how to
+log messages in general, and provide a table that outlines when to use each log
+level. The final section explains how a user can implement a custom logger.
+
+Understanding The Log Output
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+In iceoryx2, a log message typically consists of two parts: the origin and the
+actual message. The origin refers to the debug output of the object that
+emitted the log message.
+
+Including the origin is important because a backtrace, which points to a
+specific line in the code, is often not enough. A typical setup might instantiate
+many objects of the same type in different states, and a bug could be caused by
+a subtle detail that affects a later statement.
+
+Therefore, the full object state is added as the origin. In the following example,
+we see the log output of the `FileBuilder`, the class responsible for constructing
+a `File` object. By including the object's state, we gain valuable details about
+the file that was opened for reading.
+
+.. code-block::
+
+   +--------- monotonic counter
+   |  +------ log level (T = Trace, D = Debug, I = Info, W = Warn, E = Error, F = Fatal)
+   |  |  +--- origin: the state (debug output) of the object that emitted the log
+   |  |  |            message
+   |  |  |                                the actual log message ----------------+
+   |  |  |                                                                       |
+   |  |  |                                                                       |
+   0 [T] FileBuilder { file_path: FilePath { value: FixedSizeByteString<255> {   |
+         len: 20, data: "config/iceoryx2.toml" } }, access_mode: Read,           |
+         permission: Permission(448), has_ownership: false, owner: None,         |
+         group: None, truncate_size: None, creation_mode: None }                 |
+         | opened ---------------------------------------------------------------+
```

Comment:
Not in this PR but would it be possible to have the actual log message before the origin?

I volunteer to change this, if you agree

Reply:
Why? For me it feels wrong but if @orecham also prefers to switch the arguments I would go with the flow.

Reply:
Because it feels like yoda notation :sweat_smile: 

At first it's the log level, e.g. `[W]`. Then it's the context, but for what? Finally, it's the actual log message and the context becomes meaningful but one has to scroll back to the top.

I would even change the format to this in order to have a clean separation between log message and context
```
0 [T] opened
      === Context ===
      FileBuilder { file_path: FilePath { value: FixedSizeByteString<255> {   |
      len: 20, data: "config/iceoryx2.toml" } }, access_mode: Read,           |
      permission: Permission(448), has_ownership: false, owner: None,         |
      group: None, truncate_size: None, creation_mode: None } 
```

Instead of `### Context ###` something like `========` would also work.